fdbserver/ConsistencyScan.actor.cpp

```diff
@@ -0,0 +1,1090 @@
+
+				// Check that each shard has the same data on all storage servers that it resides on
+				wait(::success(checkDataConsistency(self->db,
+				                                    keyLocations,
+				                                    self->configuration,
+				                                    tssMapping,
+				                                    false /* quiescentCheck */,
+				                                    false /* tssCheck */,
+				                                    true /* firstClient */,
+				                                    false /* failureIsError */,
+				                                    0 /* clientId */,
+				                                    1 /* clientCount */,
+				                                    false /* distributed */,
+				                                    false /* shuffleShards */,
+				                                    1 /* shardSampleFactor */,
+				                                    deterministicRandom()->randomInt64(0, 10000000),
+				                                    self->finishedRounds /* repetitions */,
+				                                    &(self->bytesReadInPrevRound),
+				                                    self->restart,
+				                                    self->maxRate,
+				                                    self->targetInterval,
+				                                    self->progressKey)));
+			}
+		}
+	} catch (Error& e) {
+		if (e.code() == error_code_transaction_too_old || e.code() == error_code_future_version ||
+		    e.code() == error_code_wrong_shard_server || e.code() == error_code_all_alternatives_failed ||
+		    e.code() == error_code_process_behind || e.code() == error_code_actor_cancelled)
+			TraceEvent("ConsistencyScan_Retry").error(e); // FIXME: consistency check does not retry in this case
+		else
+			throw;
+	}
+
+	TraceEvent("ConsistencyScan_FinishedCheck");
+
+	//  Update the ConsistencyScanInfo object and persist to the database
+	csInfo.last_round_finish = StorageMetadataType::currentTime();
+	csInfo.finished_rounds = self->finishedRounds + 1;
+	auto duration = csInfo.last_round_finish - csInfo.last_round_start;
+	csInfo.smoothed_round_duration.setTotal((double)duration);
+	csInfo.progress_key = self->progressKey;
+	csInfo.bytes_read_prev_round = self->bytesReadInPrevRound;
+	loop {
+		try {
+			tr->reset();
+			tr->setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+			wait(ConsistencyScanInfo::setInfo(tr, csInfo));
+			wait(tr->commit());
+			break;
+		} catch (Error& e) {
+			wait(tr->onError(e));
+		}
+	}
+
+	return Void();
+}
+
+ACTOR Future<Void> watchConsistencyScanInfoKey(ConsistencyScanData* self) {
+	state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(self->db);
+
+	loop {
+		try {
+			tr->reset();
+			tr->setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+
+			state Optional<Value> val = wait(ConsistencyScanInfo::getInfo(tr));
+			if (val.present()) {
```

Comment:
Could ConsistencyScanData just hold a ConsistencyScanInfo object to use as its state?  Seems many of the members are the same.

Reply:
Yeah I think it could be done. But given the time left, I am reluctant to make too many changes now. I prioritized addressing your other comments that seemed more urgent. I will leave a TODO in the comments to revisit later.